maro/rl/algorithms/dqn.py

```diff
@@ -0,0 +1,122 @@
+
+
+class DQN(AbsAlgorithm):
+    """The Deep-Q-Networks algorithm.
+
+    See https://web.stanford.edu/class/psych209/Readings/MnihEtAlHassibis15NatureControlDeepRL.pdf for details.
+
+    Args:
+        eval_model (nn.Module): trainable Q-value model for computing actions given states.
+        optimizer_cls: torch optimizer class for the eval model. If this is None, the eval model is not trainable.
+        optimizer_params: parameters required for the eval optimizer class.
+        loss_func (Callable): loss function for the value model.
+        hyper_params: hyper-parameter set for the DQN algorithm.
+        target_model (nn.Module): Q-value model to train the ``eval_model`` against and to be updated periodically. If
+            it is None, the target model will be initialized as a deep copy of the eval model.
+    """
+    def __init__(
+        self, eval_model: nn.Module, optimizer_cls, optimizer_params, loss_func, hyper_params: DQNHyperParams,
+        target_model=None
+    ):
+        super().__init__()
+        self._device = torch.device("cuda" if torch.cuda.is_available() else "cpu")
+        self._model_dict = {"eval": eval_model.to(self._device)}
+        if optimizer_cls is not None:
+            self._optimizer = optimizer_cls(self._model_dict["eval"].parameters(), **optimizer_params)
+            if target_model is None:
+                self._model_dict["target"] = clone(eval_model).to(self._device) if eval_model is not None else None
+            else:
+                self._model_dict["target"] = target_model.to(self._device)
+        self._loss_func = loss_func
+        self._hyper_params = hyper_params
+        self._train_cnt = 0
+
+    @property
+    def eval_model(self):
+        return self._model_dict["eval"]
+
+    def choose_action(self, state: np.ndarray, epsilon: float = None):
+        if epsilon is None or np.random.rand() > epsilon:
+            state = torch.from_numpy(state).unsqueeze(0)
+            self._model_dict["eval"].eval()
+            with torch.no_grad():
+                q_values = self._model_dict["eval"](state)
+            return q_values.argmax(dim=1).item()
+
+        return np.random.choice(self._hyper_params.num_actions)
+
+    def train(self, states: np.ndarray, actions: np.ndarray, rewards: np.ndarray, next_states: np.ndarray):
+        if hasattr(self, "_optimizer"):
+            states = torch.from_numpy(states).to(self._device)  # (N, state_dim)
+            actions = torch.from_numpy(actions).to(self._device)  # (N,)
+            rewards = torch.from_numpy(rewards).to(self._device)   # (N,)
+            next_states = torch.from_numpy(next_states).to(self._device)  # (N, state_dim)
+            if len(actions.shape) == 1:
+                actions = actions.unsqueeze(1)   # (N, 1)
+            current_q_values = self._model_dict["eval"](states).gather(1, actions).squeeze(1)   # (N,)
+            next_q_values = self._model_dict["target"](next_states).max(dim=1)[0]   # (N,)
```

Comment:
with torch.no_grad()

Reply:
I forgot to add this in __init__:
 
for param in self._model_dict["target"].parameters():
     param.requires_grad = False